auth/client/src/main/java/org/wildfly/security/auth/client/AuthenticationConfiguration.java

```diff
@@ -679,6 +684,70 @@ public AuthenticationConfiguration usePassword(String password) {
         return usePassword(password == null ? null : ClearPassword.createRaw(ClearPassword.ALGORITHM_CLEAR, password.toCharArray()));
     }
 
+    /**
+     * Create a new configuration which is the same as this configuration, but converts the given masked password to a
+     * clear password and uses the clear password to authenticate.
+     *
+     * @param password the password to use
+     * @return the new configuration
+     */
+    public AuthenticationConfiguration useMaskedPassword(MaskedPassword password) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        Assert.assertNotNull(password);
+        final PasswordFactory passwordFactory = PasswordFactory.getInstance(password.getAlgorithm());
+        final ClearPasswordSpec spec = passwordFactory.getKeySpec(password, ClearPasswordSpec.class);
+        final char[] clearPassword = spec.getEncodedPassword();
+
+        PasswordFactory factory = PasswordFactory.getInstance(ClearPassword.ALGORITHM_CLEAR);
+        Password finalPassword = factory.generatePassword(new ClearPasswordSpec(clearPassword)).castAs(ClearPassword.class);
+        final CredentialSource filtered = getCredentialSource().without(PasswordCredential.class);
+        return finalPassword == null ? useCredentials(filtered) : useCredentials(filtered).useCredential(new PasswordCredential(finalPassword));
+    }
+
+    /**
+     * Create a new configuration which is the same as this configuration, but which uses the given masked password to authenticate.
+     *
+     * @param maskedPasswordBytes the masked password bytes (must not be {@code null})
+     * @param algorithm the algorithm (can be {@code null}, default:"masked-MD5-DES")
+     * @param initialKeyMaterial the initial key material (can be {@code null}, default:"somearbitrarycrazystringthatdoesnotmatter")
+     * @param iterationCount the iteration count (must not be less than 1)
+     * @param salt the salt bytes (must not be {@code null})
+     * @param initializationVector the initialization vector (can be {@code null})
+     * @return the new configuration
+     */
+    public AuthenticationConfiguration useMaskedPassword(byte[] maskedPasswordBytes, String algorithm, char[] initialKeyMaterial, int iterationCount, byte[] salt, byte[] initializationVector) throws Exception {
```

Comment:
It would be good to indicate the specific exceptions that are thrown (instead of using ```throws Exception```). It would also be good to add an ```@throws``` line to the javadoc above.

Reply:
Fixed